api/src/main/java/org/openmrs/util/DatabaseUtil.java

```diff
@@ -94,9 +93,80 @@ public static String loadDatabaseDriver(String connectionUrl, String connectionD
 	}
 	
 	/**
-	 * Executes the passed SQL query, enforcing select only if that parameter is set
+	 * Executes the passed SQL query, enforcing select only if that parameter is set for given Session
 	 */
-	public static List<List<Object>> executeSQL(Connection conn, String sql, boolean selectOnly) throws DAOException {
+	public static List<List<Object>> executeSQLWithSession(Session session, String sql, boolean selectOnly)
+	        throws DAOException {
+		sql = sql.trim();
+		boolean dataManipulation = false;
+		
+		String sqlLower = sql.toLowerCase();
+		if (sqlLower.startsWith("insert") || sqlLower.startsWith("update") || sqlLower.startsWith("delete")
+		        || sqlLower.startsWith("alter") || sqlLower.startsWith("drop") || sqlLower.startsWith("create")
+		        || sqlLower.startsWith("rename")) {
+			dataManipulation = true;
+		}
+		
+		if (selectOnly && dataManipulation) {
+			throw new IllegalArgumentException("Illegal command(s) found in query string");
+		}
+		
+		final List<List<Object>> result = new ArrayList<List<Object>>();
+		final String query = sql;
+		final boolean sessionDataManipulation = dataManipulation;
+		
+		//todo replace with lambdas after moving on to Java 8
+		session.doWork(new Work() {
+			
+			@Override
+			public void execute(Connection conn) {
+				PreparedStatement ps = null;
```

Comment:
Instead of duplicating code you should call "executeSQL(connection, sql, selectOnly)". Use session.doReturningWork to free yourself from using the final result object.